File: app/src/main/java/com/chiului/android_mvvm_architecture/api/BaseApiService.java
package com.chiului.android_mvvm_architecture.api;

import android.util.Log;

import com.chiului.android_mvvm_architecture.BuildConfig;
import com.chiului.android_mvvm_architecture.utilities.Constants;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static android.content.ContentValues.TAG;

/**
 * 网络请求基础服务$
 *
 * @author 神经大条蕾弟
 * @date 2020/08/31 15:26
 */
public class BaseApiService {

    private static Retrofit mRetrofit;

    /**
     * 获取Retrofit
     *
     * @return 获取Retrofit
     */
    public static Retrofit getRetrofit() {
        if (mRetrofit == null) {

            HttpLoggingInterceptor logger = new HttpLoggingInterceptor(message -> {
                if (BuildConfig.DEBUG) {
                    Log.e(TAG, message);
                }
            });
            logger.setLevel(HttpLoggingInterceptor.Level.BODY);// 设置打印级别
            logger.redactHeader("Authorization");// 删除敏感信息
            logger.redactHeader("Cookie");// 删除敏感信息

            OkHttpClient client = new OkHttpClient.Builder()
                    .addInterceptor(logger)
                    .build();

            Retrofit.Builder builder = new Retrofit.Builder()
                    .baseUrl(Constants.AIP_BASE)// 配置BaseUrl
                    .client(client)// 设置client
                    .addConverterFactory(GsonConverterFactory.create())// gson转换器
                    .addCallAdapterFactory(RxJavaCallAdapterFactory.create());// RxJava 支持
            mRetrofit = builder.build();
        }
        return mRetrofit;
    }

}

File: app/src/main/java/com/chiului/android_mvvm_architecture/api/service/BaseApiService.kt
package com.chiului.android_mvvm_architecture.api.service

import com.chiului.android_mvvm_architecture.BuildConfig
import com.chiului.android_mvvm_architecture.MainApplication
import com.chiului.android_mvvm_architecture.api.converter.CustomGsonConverterFactory
import com.chiului.android_mvvm_architecture.data.AppDatabase
import com.chiului.android_mvvm_architecture.utilities.AIP_BASE
import com.chiului.android_mvvm_architecture.utilities.TOKEN
import com.chiului.android_mvvm_architecture.utilities.logI
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory

/**
 * 获取 Retorfit Service$
 * @author    神经大条蕾弟
 * @date      2021/01/14 15:43
 */
object BaseApiService {

    fun getRetrofit(): Retrofit{

        // 日志拦截器
        val loggingInterceptor = HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
            override fun log(message: String) {
                if (BuildConfig.DEBUG) {
                    // 打印
                    message.logI()
                }
            }
        }).apply {
            level = HttpLoggingInterceptor.Level.BASIC // 设置打印级别
            redactHeader("Authorization")// 删除敏感信息
            redactHeader("Cookie")// 删除敏感信息
        }


        // 请求头拦截器
        val headerInterceptor = Interceptor { chain: Interceptor.Chain ->
            val original = chain.request()
            val requestBuilder: Request.Builder = original.newBuilder()
                    .header("token", getToken()) // 添加或替换 token 到请求头
            val request: Request = requestBuilder.build()
            chain.proceed(request)
        }

        val client: OkHttpClient = OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .addInterceptor(headerInterceptor)
                .build()

        val builder: Retrofit.Builder = Retrofit.Builder()
                .baseUrl(AIP_BASE) // 配置BaseUrl
                .client(client) // 设置client
//                .addConverterFactory(GsonConverterFactory.create()) // gson转换器
                .addConverterFactory(CustomGsonConverterFactory.create()) // 自定义 gson 转换器
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create()) // RxJava3 支持

        return builder.build()
    }

    private fun getToken(): String {
        var appCache = AppDatabase.getInstance(MainApplication.context).appCacheDao().getAppCache(TOKEN)
        return if(appCache == null) "" else appCache.value ?: ""
    }

    @JvmStatic
    fun getJsonRequestBody(json: String): RequestBody {
        return json.toRequestBody("application/json;charset=UTF-8".toMediaType())
    }

}